src/hooks/use-load-data-pages.ts

```diff
@@ -55,17 +55,19 @@ export const isMissingLayoutData = (
  * guaranteed to work with a new layout, as its properties may no longer be valid.
  */
 export const getFetchArea = (
-  qLastExpandedPos: EngineAPI.INxCellPosition | undefined,
+  triggerdByExpandOrCollapse: boolean,
   viewService: ViewService,
   qSize: EngineAPI.ISize,
   pageInfo: PageInfo,
   maxNumberOfVisibleRows: number,
   maxNumberOfVisibleColumns: number,
 ) => {
-  const pageStartIndex = pageInfo.page * pageInfo.rowsPerPage;
-  const pageEndIndex = pageStartIndex + pageInfo.rowsPerPage;
+  const pageRowStartIndex = pageInfo.page * pageInfo.rowsPerPage;
+  // Do not fetch data beyond this value. Either because there is no more data in the layout or the current page ends.
+  const rowEndIndex = Math.min(pageRowStartIndex + pageInfo.rowsPerPage, qSize.qcy);
+  const columnEndIndex = Math.min(MAX_COLUMN_COUNT, qSize.qcx);
```

Comment:
let's imagine a scenario where we have 50k cols in dataset
`MAX_COLUMN_COUNT` is 15k and qcx shows us the above 50k
does this means when we fetch new batch of data, we will always fetch batches in 15k size in width?

probably I'm missing something but as far as i checked with line 153 in this module, seems like we just subtracting what ever current `qLeft` is from  `columnEndIndex`, not sure if `maxNumberOfVisibleColumns` will cover that scenario? 

In the other hand, seems like `maxNumberOfVisibleColumns` gives us some bigger number than we expect, for example, there is like 10 cells visible in dataGrid below, but we get 42 cells as `maxNumberOfVisibleColumns`: 

<img width="2376" alt="Screenshot 2023-12-04 at 11 16 49" src="https://github.com/qlik-oss/sn-pivot-table/assets/29652890/fe017edb-286b-4e7f-b1ef-6b23f8b4209b">

a bit of digging into this  i notices the initial width of rect.width:

<img width="1909" alt="Screenshot 2023-12-04 at 11 18 14" src="https://github.com/qlik-oss/sn-pivot-table/assets/29652890/8b1851b3-4496-41bd-a625-dcff1d1c4db5">

and here is the calculation from `handle-max-engine-size`:

<img width="1279" alt="Screenshot 2023-12-04 at 11 20 20" src="https://github.com/qlik-oss/sn-pivot-table/assets/29652890/235755dc-6539-4231-aec5-8b01251a631b">
 
seems like it considers the cell min pixel values and this means for every fetch -> we will endup fetching some cells more than expected?
I mean, AFAIunderstood the code, from first part of my comment -> we are rely on `maxNumberOfVisibleColumns` when it comes to fetching a new batch of data (always, but consider the large col scenario like 50k for now), and meanwhile this way, `maxNumberOfVisibleColumns` calculated based on min cell pixels size which returns more cell counts than what user can see (like without subtracting column sizes from left grid, and not the actual size of rendered cell in data grid -> because visually what users can see is ~10 to 12 cells)

Reply:
Yes, absolutely!
the thing that i was thinking was, when we taking the safe guess amount for cells, we dont subtract the left grid width from rect.width, but as you said at that point in render flow, we dont have access to left grid element to actually measure its width and subtract it fomr react.width, so it looks fine if we consider to fetch more cells than the real actual width is. 

one last concern could be this below scenario:

<img width="1960" alt="Screenshot 2023-12-04 at 11 56 54" src="https://github.com/qlik-oss/sn-pivot-table/assets/29652890/83629aa4-0641-4a77-bd79-adc09934cb08">

where we could have a large data cells (for some reason), and of course large value for react.width (4k, 8k size), 
in that case the cells we will fetch is going to be 8194/30 =  ~273 ish, but user may see much less cells than that. 

That is like more unlikely edge case, but not sure if it's worth the effort + that much important to take care of.

Reply:
Yes there are scenarios where we end up fetching way more then needed. On an 8k resolution we will fetch ~37 000 cells, but many of them are likely not going to be visisble.

It would be nice to have some metrics on the number of rows/columns the user actually have in their data, so that we can fine-tune our solution around that.